src/main/java/com/powsybl/openloadflow/ac/SecondaryVoltageControlOuterLoop.java

```diff
@@ -0,0 +1,323 @@
+/**
+ * Copyright (c) 2023, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.openloadflow.ac;
+
+import com.powsybl.commons.reporter.Reporter;
+import com.powsybl.math.matrix.DenseMatrix;
+import com.powsybl.openloadflow.ac.equations.AcEquationType;
+import com.powsybl.openloadflow.ac.equations.AcVariableType;
+import com.powsybl.openloadflow.ac.outerloop.AcLoadFlowContext;
+import com.powsybl.openloadflow.ac.outerloop.OuterLoop;
+import com.powsybl.openloadflow.ac.outerloop.OuterLoopContext;
+import com.powsybl.openloadflow.ac.outerloop.OuterLoopStatus;
+import com.powsybl.openloadflow.equations.EquationSystem;
+import com.powsybl.openloadflow.equations.EquationTerm;
+import com.powsybl.openloadflow.equations.JacobianMatrix;
+import com.powsybl.openloadflow.network.*;
+import com.powsybl.openloadflow.util.PerUnit;
+import org.apache.commons.lang3.mutable.MutableDouble;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class SecondaryVoltageControlOuterLoop implements OuterLoop {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SecondaryVoltageControlOuterLoop.class);
+
+    private static final double TARGET_V_DIFF_EPS = 1e-1; // in Kv
+    private static final double SENSI_V_EPS = 1e-3;
+
+    @Override
+    public String getType() {
+        return "SecondaryVoltageControl";
+    }
+
+    private static boolean isValid(LfBus bus) {
```

Comment:
Why these three methods `isValid`, `getControllerBuses` and `findActiveSecondaryVoltageControls` are not methods of objects `LfBus` and `LfNetwork`?

Reply:
Because filtering criteria is too specific to this outer loop. We should avoid moving to specific code to LfNetwork.